File: common/src/main/java/com/ssblur/scriptor/helpers/targetable/ContainerTargetable.java
package com.ssblur.scriptor.helpers.targetable;

import net.minecraft.core.BlockPos;
import net.minecraft.world.Container;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.Nullable;

public class ContainerTargetable extends Targetable implements InventoryTargetable {
  int slot;
  public ContainerTargetable(Level level, BlockPos pos, int slot) {
    super(level, pos);
    this.slot = slot;
  }

  @Override
  public @Nullable Container getContainer() {
    if(level.getBlockEntity(getTargetBlockPos()) instanceof Container container)
      return container;
    return null;
  }

  @Override
  public int getTargetedSlot() {
    return slot;
  }

  @Override
  public void setTargetedSlot(int slot) {
    this.slot = slot;
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/helpers/targetable/ContainerTargetable.kt
package com.ssblur.scriptor.helpers.targetable

import com.ssblur.scriptor.mixin.BaseContainerBlockEntityAccessor
import net.minecraft.core.BlockPos
import net.minecraft.world.Container
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BaseContainerBlockEntity

class ContainerTargetable(level: Level, pos: BlockPos, var slot: Int): Targetable(level, pos), InventoryTargetable {
  override val container: Container?
    get() {
      val blockEntity = level.getBlockEntity(targetBlockPos)
      if (blockEntity is Container) {
        if (blockEntity is BaseContainerBlockEntity) {
          if ((blockEntity as BaseContainerBlockEntityAccessor).lockKey.unlocksWith(ItemStack.EMPTY))
            return level.getBlockEntity(targetBlockPos) as Container
          return null
        }
        return level.getBlockEntity(targetBlockPos) as Container
      }
      return null
    }

  override var targetedSlot = slot
}
